src/foundry/common/utils/iterable-weak-map.d.mts

```diff
@@ -1,9 +1,85 @@
-export default IterableWeakMap;
+interface IterableWeakMapHeldValue<K extends WeakKey> {
+  /**
+   * The set to be cleaned.
+   */
+  set: Set<WeakRef<K>>;
+
+  /**
+   * The ref to remove.
+   */
+  ref: WeakRef<K>;
+}
 
 /**
  * Stores a map of objects with weak references to the keys, allowing them to be garbage collected. Both keys and values
  * can be iterated over, unlike a WeakMap.
+ * @typeParam K - the type of the keys contained in the IterableWeakMap
+ * @typeParam V - The type of the values contained in the IterableWeakMap
  */
-declare class IterableWeakMap extends WeakMap {
-  // TODO: Define methods and properties
+declare class IterableWeakMap<K extends WeakKey, V extends object> extends WeakMap<K, V> {
+  /**
+   * A set of weak refs to the map's keys, allowing enumeration.
+   */
+  #refs: Set<WeakRef<K>>;
+
+  /**
+   * A FinalizationRegistry instance to clean up the ref set when objects are garbage collected.
+   */
+  #finalizer: FinalizationRegistry<IterableWeakMapHeldValue<K>>;
+
+  /**
+   * @param entries - The initial entries.
+   */
+  constructor(entries?: Iterable<[K, V]>);
+
+  /**
+   * Clean up the corresponding ref in the set when its value is garbage collected.
+   * @param heldValue - The value held by the finalizer.
+   */
+  static #cleanup<K extends WeakKey>(heldValue: IterableWeakMapHeldValue<K>): void;
+
+  /**
+   * Remove a key from the map.
+   * @param key - The key to remove.
+   * @returns `true` if a key was removed, `false` otherwise.
+   */
+  delete(key: K): boolean;
+
+  /**
+   * Retrieve a value from the map.
+   * @param key - The value's key.
+   * @returns
+   */
+  get(key: K): V | undefined;
+
+  /**
+   * Place a value in the map.
+   * @param key   - The key.
+   * @param value - The value.
+   * @returns
+   */
+  set(key: K, value: V): this;
+
+  /**
+   * Enumerate the entries.
+   * @returns
+   */
+  [Symbol.iterator](): Generator<[K, V], void, any>;
```

Comment:
The first two parameters look fine but I'd probably make the third `never`.

```
declare const someGenerator: Generator<[K, V], void, any>;

// calling `next` this way doesn't really make sense
someGenerator.next(1);
someGenerator.next("42");
someGenerator.next(new class {});

declare const withNever: Generator<[K, V], void, never>;

// All errors
withNever.next(1);
withNever.next("42");
withNever.next(new class {});
```

See https://tsplay.dev/w8qMEN

Reply:
That makes a lot of sense. Thanks for the example.